Mathlib/CategoryTheory/Comma/Final.lean

```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2024 Jakob von Raumer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob von Raumer
+-/
+import Mathlib.CategoryTheory.Functor.KanExtension.Adjunction
+import Mathlib.CategoryTheory.Limits.Final
+import Mathlib.CategoryTheory.Grothendieck
+
+/-!
+# Finality of Projections in Comma Categories
+-/
+
+universe v₁ v₂ v₃ u₁ u₂ u₃
+
+namespace CategoryTheory
+
+namespace Comma
+
+open Limits Functor CostructuredArrow
+
+section Small
+
+variable {A : Type v₁} [Category.{v₁} A]
+variable {B : Type v₁} [Category.{v₁} B]
+variable {T : Type v₁} [Category.{v₁} T]
+variable (L : A ⥤ T) (R : B ⥤ T)
+
+private lemma final_fst_small [R.Final] : (fst L R).Final := by
+  rw  [Functor.final_iff_isIso_colimit_pre]
+  intro G
+  let i : colimit G ≅ colimit (fst L R ⋙ G) :=
+    colimitIsoColimitGrothendieck L G ≪≫
+    (Final.colimitIso (Grothendieck.pre (functor L) R) (grothendieckProj L ⋙ G)).symm ≪≫
+    HasColimit.isoOfNatIso (Iso.refl _) ≪≫
+    Final.colimitIso (grothendieckPrecompFunctorEquivalence L R).functor (fst L R ⋙ G)
+  convert i.isIso_inv
+  apply colimit.hom_ext
+  intro ⟨a, b, f⟩
+  simp only [colimit.ι_pre, comp_obj, fst_obj, grothendieckPrecompFunctorEquivalence_functor,
+    Iso.trans_inv, Iso.symm_inv, Category.assoc, i]
+  change _ = colimit.ι (fst L R ⋙ G)
+    ((grothendieckPrecompFunctorToComma L R).obj ⟨b, CostructuredArrow.mk f⟩) ≫ _
+  simp
+
+end Small
+
+variable {A : Type u₁} [Category.{v₁} A]
+variable {B : Type u₂} [Category.{v₂} B]
+variable {T : Type v₃} [Category.{v₃} T]
+variable (L : A ⥤ T) (R : B ⥤ T)
+
+lemma final_fst [R.Final] : (fst L R).Final := by
```

Comment:
I'm surprised that we don't seem to have
```lean
import Mathlib

universe v₁ v₂ v₃ u₁ u₂ u₃

open Opposite

namespace CategoryTheory

variable {A : Type v₁} [Category.{v₁} A]
variable {B : Type v₁} [Category.{v₁} B]
variable {T : Type v₁} [Category.{v₁} T]
variable (L : A ⥤ T) (R : B ⥤ T)

@[simps]
def Comma.opFunctor : Comma L R ⥤ (Comma R.op L.op)ᵒᵖ where
  obj X := ⟨op X.right, op X.left, op X.hom⟩ 
  map f := ⟨op f.right, op f.left, Quiver.Hom.unop_inj (by simp)⟩

@[simps!]
def Comma.opFunctorCompFst :
    Comma.opFunctor L R ⋙ (Comma.fst _ _).op ≅
      Comma.snd _ _ ⋙ (opOpEquivalence _).inverse :=
  Iso.refl _

@[simps!]
def Comma.opFunctorCompSnd :
    Comma.opFunctor L R ⋙ (Comma.snd _ _).op ≅
      Comma.fst _ _ ⋙ (opOpEquivalence _).inverse :=
  Iso.refl _

@[simps]
def Comma.unopFunctor : Comma L.op R.op ⥤ (Comma R L)ᵒᵖ where
  obj X := ⟨X.right.unop, X.left.unop, X.hom.unop⟩ 
  map f := ⟨f.right.unop, f.left.unop, Quiver.Hom.op_inj (by simpa using f.w.symm)⟩

@[simps!]
def Comma.unopFunctorCompFst : Comma.unopFunctor L R ⋙ (Comma.fst _ _).op ≅ Comma.snd _ _ :=
  Iso.refl _

@[simps!]
def Comma.unopFunctorCompSnd : Comma.unopFunctor L R ⋙ (Comma.snd _ _).op ≅ Comma.fst _ _ :=
  Iso.refl _

@[simps]
def Comma.opEquiv : Comma L R ≌ (Comma R.op L.op)ᵒᵖ where
  functor := Comma.opFunctor L R
  inverse := (Comma.unopFunctor R L).leftOp
  unitIso := NatIso.ofComponents (fun X => Iso.refl _)
  counitIso := NatIso.ofComponents (fun X => Iso.refl _)

end CategoryTheory
```
but with this it should hopefully be able to deduce the dual statement about initial functors?